gcl_iam/controllers.py

```diff
@@ -0,0 +1,169 @@
+#
+#    All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+import uuid
+
+from restalchemy.api import controllers
+from restalchemy.common import contexts
+from restalchemy.dm import types
+
+from gcl_iam import enforcers
+from gcl_iam import exceptions
+
+
+class PolicyBasedControllerMixin(object):
+
+    __policy_name__ = None
+
+    def __init__(self, *args, **kwargs):
+        self._introspection = (
+            contexts.get_context().iam_context.introspection_info()
+        )
+
+        # Forbid requests without auth or without project scope
+        if not self._introspection:
+            raise exceptions.Unauthorized()
+
+        self._ctx_project_id = self._introspection.get("project_id", None)
+        self._enforcer = enforcers.Enforcer(self._introspection["permissions"])
+
+        super().__init__(*args, **kwargs)
+
+    def _enforce(self, method):
+        if self.__policy_name__:
+            policy_key = (
+                "genesis_core." + self.__policy_name__ + "." + method
+            )  # TODO
+        else:
+            policy_key = "default"
+
+        return self._enforcer.enforce(policy_key, do_raise=True)
+
+    def _force_project_id(self, project_id):
+        # ghosts converts project ids to true UUIDs with dashes
+        if not isinstance(project_id, uuid.UUID):
+            project_id = uuid.UUID(project_id)
+        if project_id != self._ctx_project_id:
+            raise exceptions.Forbidden()
+
+    def _enforce_and_authorize_project_id(self, method, project_id):
+        if self._enforce(method) and not self._ctx_project_id:
+            return
+
+        self._force_project_id(project_id)
+
+    def _enforce_and_override_project_id_in_kwargs(self, method, kwargs):
+        if self._enforce(method) and not self._ctx_project_id:
+            return
+
+        if "project_id" in kwargs:
+            self._force_project_id(kwargs["project_id"])
+        else:
+            kwargs["project_id"] = types.UUID().from_simple_type(
+                self._ctx_project_id
+            )
+
+
+class PolicyBasedController(
+    PolicyBasedControllerMixin, controllers.BaseResourceController
+):
+
+    def create(self, **kwargs):
+        self._enforce_and_override_project_id_in_kwargs("create", kwargs)
+
+        return super(PolicyBasedControllerMixin, self).create(**kwargs)
+
+    def get(self, **kwargs):
+        right = self._enforce("get")
```

Comment:
Properly speaking, we should incorporate the project into the filters here. The user should receive a 404 error if the project is incorrect. It might be worth refining the common controller: move the logic for generating filters for all requests into a shared method that can be overridden.

Reply:
Done